odb_extractor_atoms.py

```diff
@@ -445,340 +878,157 @@ def xlsxseq(file, path):
-    # file = 'GN-2018B-DD-104.json'
-    # with open(os.path.join(path, file), 'r') as f:
-    #     program = json.load(f)
-    #
-    #
-    # for prog in list(program.keys()):
-    # #     print(list(program[prog].keys()))
-    #     for item in list(program[prog].keys()):
-    #         print(item)
-    # #         print(program[prog][item])
-    #         if 'INFO' in item:
-    #             print(f"\t {program[prog][item]['title']}")
-    #         if all(type not in item for type in ['INFO', 'GROUP']):
-    #             print(f" \t {program[prog][item]}")
-    #
-    #
-    # prog = list(program.keys())[0]
-    # group = 'GROUP_GROUP_SCHEDULING-13'
-    # obsnum = []
-    # for item in list(program[prog][group].keys()):
-    #     obsid = ''
-    #     if 'OBSERVATION' in item:
-    #         obsid = program[prog][group][item]['sequence'][0]['ocs:observationId']
-    #         obsnum.append(int(item.split('-')[1]))
-    #         print(item, obsnum[-1], obsid)
-    #     else:
-    #         print(item, program[prog][group][item])
-    #
-    # obsnum.sort()
-    # # print(obsnum)
-    # for ii in obsnum:
-    #     item = 'OBSERVATION_BASIC-' + str(ii)
-    #     obsid = program[prog][group][item]['sequence'][0]['ocs:observationId']
-    #     print(ii, obsid, program[prog][group][item]['sequence'][0]['instrument:instrument'])
-    #
-    # obs = 'OBSERVATION_BASIC-7'
-    # for item in list(program[prog][group][obs].keys()):
-    #     print(item)
-    #     if 'INFO' in item:
-    #             print(f"\t {program[prog][group][obs][item]['title']}")
-    #     if any(type in item for type in ['CONDITIONS', 'TARGETENV']):
-    #         for key in list(program[prog][group][obs][item].keys()):
-    #             print(f"\t {key}")
-    #     if all(type not in item for type in ['INFO', 'sequence', 'obsLog']):
-    #             print(f" \t {program[prog][group][obs][item]}")
+            # New atom?
+            if nextatom:
+                # Get class, qastate, guiding for previous atom
+                if natom > 0:
+                    atoms[-1]['qa_state'] = select_qastate(qastates)
+                    if atoms[-1]['qa_state'] != 'NONE':
+                        atoms[-1]['observed'] = True
+                    atoms[-1]['class'] = select_obsclass(classes)
+                    if verbose:
+                        print('QA states: ', qastates)
+                        print('Classes: ', classes)
+                        print('Guiding: ', guiding)
+                    atoms[-1]['guide_state'] = any(guiding)
+                    atoms[-1]['wavelength'] = sequence['wavelength'][-1]
+                    atoms[-1]['required_resources']['inst'] = sequence['inst'][-1]
+                    atoms[-1]['required_resources']['filter'] = sequence['filter'][-1]
+                    atoms[-1]['required_resources']['disperser'] = sequence['disperser'][-1]
+                    atoms[-1]['required_resources']['fpu'] = sequence['fpu'][-1]
```

Comment:
This pattern seems to be coming up repeatedly, which suggests to me that we need to factor it out.